### ISSUE TITLE
Maximum space used by snaps, feature request

### ISSUE THREAD

**Issue by zeigerpuppy** [jimsalterjrs/sanoid#337]
It would be useful have an (optional) limit on the space used by snapshots.  This is of benefit because quotas in ZFS include the snapshot data space (potentially leading to snapshots 'stealing' available space from a filesystem).

This would permit reserving space for the filesystem of a dataset.

For instance, if one wants 50GB available and 20GB for snapshots, then the following could be set:

```bash
zfs set quota=70G pool/dataset
```

potential change to sanoid config:

```
[data/images]
	use_template = hardlimit
	recursive = yes
	process_children_only = yes

[template_hardlimit]
        frequently = 0
        hourly = 36
        daily = 30
        monthly = 3
        yearly = 0
        autosnap = yes
        autoprune = yes
        spacelimit = 20G
        spaceprune = oldest
```

then there would always be 50GB available to the filesystem
`spaceprune` could have optional algorithms for smart pruning, e.g. `oldest|largest|prompt`, which could be:
- `oldest`: deletes oldest snaps first to get space back under quota (if the next snap fails, it deletes more)
- `largest`: deletes largest snap first
- `prompt`: prompts user (by email) when snaps exceed the `spacelimit` 

I haven't thought through the algorithm too much but the space occupied by snaps is pretty easy to access:

```
zfs list -ro space pool/dataset
NAME  AVAIL   USED  USEDSNAP  USEDDS  USEDREFRESERV  USEDCHILD
 pool/dataset  70.0G  46.3M     23.5M   22.8M             0B         0B
```

**Comment by apiening** [comment]
Just my thoughts on this: The size of the snapshot is not static. The snapshot grows when data is changed on the volume.
For example if you choose spacelimit = 20G and there is 25G deleted in the volume, the snapshot would grow more then 20G, right?
How would sanoid even be able to notice that? It could just check the size of the snapshots when it is called the next time (for example with cron).
And if you have set spaceprune=largest, you would immediately lose everything that might have been unintentionally deleted. But even with spaceprune=oldest, sanoid would need to delete all other snapshots and even then it would hold the 25G from the snapshot in our example which would be still more than the spacelimit of 20G.
Even if this might be possible to implement, I think it would be dangerous to use and I can't think of a good use case for this option.

What I have set up is a monitoring check that notifies me if my pool gets over 90% full for example. This includes snapshots as well. If this ever happens I would take a manual look at the system to examine if I need to extend the pool or replace it so that there is sufficient disk space available.

**Comment by zeigerpuppy** [comment]
>For example if you choose spacelimit = 20G and there is 25G deleted in the volume, the snapshot would grow more then 20G, right?

Yes, that it true, which would make `spaceprune=largest` a bit risky, I agree.

However, a smart algorithm for `spaceprune=oldest` would recover space relatively safely, for example, let's say we have three snapshots on consecutive days when 10G of data has been altered on each day (let's assume 5G write, 5G erase each day).  Then you can easily end up in the situation where 30G of space is used.
Now, if we delete an intermediate snapshot, we get back half the space (at least), so deleting one snapshot gives up 5G back and deleting all three (so that snapshot 0 goes to snapshot 1) will give back at least 15GB.  
I can certainly agree, that some logic to limit damage would be very useful (for instance, don't touch the most recent n snapshots).

**Comment by redmop** [comment]
What if we did snapshots as a range?
Freq=4-8
Hourly=4-8
(etc)
Snapshots older than the range can be deleted whenever, en-mass if needed.
Snapshots within the range are deleted, say the oldest or by size in each prune pass, until  we hit the space target.
Snapshots at the min or less than the range are safe from deletion.

**Comment by zeigerpuppy** [comment]
@redmop, that sounds like a good idea, it would simplify the logic of picking the appropriate snapshots to delete

**Comment by redmop** [comment]
I don't know how much simpler it will be, big picture. It will help with picking of snapshots though.

**Comment by Kramerican** [comment]
Looking at this thread I am hoping one of you guys can clarify: if I autosnap with Sanoid, as-is this means free disk space for my datasets will continuously decrease as more snaps are made?

I'm running LXD containers on ZFS and LXd seems to use refquota (which is just quota but for a referenced dataset, I suppose?) - so if I start doing snapshots of my containers with zfs/sanoid then my users available disk space will decrease?

In that case, this feels like a fetaure that needs to be implemented in ZFS: allow snapshots to take up space independent of dataset quota.

Can any of you guys enlighten me on the ins and outs here? Thanks :)

**Comment by ghormoon** [comment]
i think what you're looking into is the difference between quota (includes snapshots and children datasets) and refquota (includes only the dataset itself)

if i understood the question right, you can achieve this directly on zfs level e.g. by setting 50g refquota and 70g quota.
similarly if you need reservation/refreservation.

is there any other usecase that needs to be supported on sanoid level that can't be done directly by the properties?
